django/db/models/query.py

```diff
@@ -815,8 +815,15 @@ def contains(self, obj):
 
         If the QuerySet is already fully cached, check if object is in cache.
         """
+        try:
+            if obj._meta.concrete_model != self.model._meta.concrete_model:
+                return False
+        except AttributeError:
+            raise NotImplementedError(
+                'QuerySet.contains only supports Model objects. You passed in a {}.'.format(type(obj))
+            )
```

Comment:
I think we should return `False` instead or raising and exception, we're not going to implement any tricky workarounds for a different objects.

Reply:
I don't agree. I anyone calls this with a non-model they definitely did something weird. It's more helpful to raise a descriptive exception than ignoring that mistake.

Reply:
If you want to raise an exception than `NotImplementedError` is not the right choice, IMO. I would use `ValueError`.

Reply:
+1 for `ValueError`